pkg/controller/controller_instance.go

```diff
@@ -1173,3 +1214,52 @@ func (c *controller) updateServiceInstanceFinalizers(
 	}
 	return err
 }
+
+func (c *controller) recordStartOfServiceInstanceOperation(toUpdate *v1alpha1.ServiceInstance, operation v1alpha1.ServiceInstanceOperation) (*v1alpha1.ServiceInstance, error) {
+	toUpdate.Status.CurrentOperation = operation
+	now := metav1.Now()
+	toUpdate.Status.OperationStartTime = &now
+	reason := ""
+	message := ""
+	switch operation {
+	case v1alpha1.ServiceInstanceOperationProvision:
+		reason = provisioningInFlightReason
+		message = provisioningInFlightMessage
+	case v1alpha1.ServiceInstanceOperationDeprovision:
+		reason = deprovisioningInFlightReason
+		message = deprovisioningInFlightMessage
+	}
+	setServiceInstanceCondition(
+		toUpdate,
+		v1alpha1.ServiceInstanceConditionReady,
+		v1alpha1.ConditionFalse,
+		reason,
+		message,
+	)
+	if err := c.updateServiceInstanceStatus(toUpdate); err != nil {
+		return toUpdate, err
+	}
+	// Get the latest version of the instance so that we can avoid
+	// conflicts on our subsequent updates
```

Comment:
What sort of conflict are we trying to avoid here? Update calls to the API server are synchronous, so we're guaranteed the API server at least accepted the change. And the worker queue is thread-safe so we wouldn't be updating the status of this instance in any other goroutine.

Reply:
Admittedly, I am not entirely sure. I was mostly following the pattern used in the updateServiceInstanceCredentialFinalizers functions, and their ilk. My thinking was that the API Server updates the resrouceVersion of the resource in storage when the call to UpdateStatus is made. This does not update the resourceVersion for the copy of the resource that we are using in the reconciliation function. If we do not get the latest version with the updated resourceVersion, then the next call to UpdateStatus in the reconciliation function will fail with a conflict because the API Server will see that we are trying to update an out-of-date Status.

Reply:
You don't need to do the get, because update returns the updated resource.

Reply:
It is not technically necessary to re-get, but given the current limitations of the fake client, I think it makes sense to keep this logic for now, with a TODO to change it once the fake client does a deep copy of the object.